### PR TITLE
Tree: Fix bugs with rebases involving cross-field moves

### PR DESCRIPTION
**Pull request by alex-pardes**
This PR fixes an issue where an edit could be incorrectly dropped when rebased over a cross-field move, as well as an issue where only one endpoint of a cross-field move was updated when rebasing it. Prior to this PR, when rebasing a change we only descended into fields which had modifications from both the new change and the base change.

This PR changes the contract of `FieldChangeHandler.rebase` to require calling `rebaseChild` on base `NodeChangeset`s even when there is no matching new change, and new `NodeChangeset`s even when there is no matching base change.

### CODE REVIEW

File: packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts [hunk 1]
```diff
@@ -442,9 +443,15 @@ export class ModularChangeFamily
 		change: ModularChangeset,
 		over: TaggedChange<ModularChangeset>,
 	): ModularChangeset {
-		let maxId = change.maxId ?? -1;
+		let maxId = Math.max(change.maxId ?? -1, over.change.maxId ?? -1);
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts:446]: Why do we need this change?

File: packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts [hunk 2]
```diff
@@ -454,127 +461,296 @@ export class ModularChangeFamily
...
 		assert(
-			crossFieldTable.fieldsToUpdate.size === 0,
+			crossFieldTable.invalidatedFields.size === 0,
 			0x59f /* Should not need more than one amend pass. */,
 		);
 
-		return makeModularChangeset(rebasedFields, maxId, change.revisions);
+		if (maxId >= 0) {
+			rebasedChangeset.maxId = brand(maxId);
+		}
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts:496]: I think that's handled by `makeModularChangeset` now.

File: packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts [hunk 3]
```diff
@@ -620,6 +796,44 @@ export class ModularChangeFamily
...
+
+		fieldsToAmend.add(fieldChange);
+	}
+
+	return [fieldsToAmend, invalidatedEmptyFields];
+}
+
+function newGenericChangeset(): GenericChangeset {
+	return [];
+}
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts:835]: Nit: move this to genericFieldKind.ts

File: packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts [hunk 4]
```diff
@@ -454,127 +461,296 @@ export class ModularChangeFamily
...
+		crossFieldTable: RebaseTable,
+		fieldsToAmend: Set<FieldChange>,
+		invalidatedEmptyFields: Set<FieldChange>,
+		genId: IdAllocator,
+		revisionMetadata: RevisionMetadataSource,
+	) {
+		crossFieldTable.invalidatedFields = new Set();
+		for (const fieldToAmend of fieldsToAmend) {
+			const fieldContext = crossFieldTable.baseChangeToContext.get(fieldToAmend);
+			let fieldChange: FieldChange;
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts:510]: Nit: consider renaming `fieldChange` to `newChange` or `newFieldChange`. The current name discribes the type, which most tooling already helps with, but does not describe which of the change changesets the `FieldChange` belongs to, which is the more relevant question here.

File: packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts [hunk 5]
```diff
@@ -454,127 +461,296 @@ export class ModularChangeFamily
...
+
+			addFieldData(manager, baseChanges);
+			crossFieldTable.baseChangeToContext.set(baseChanges, {
+				map: over.change,
+				field,
+				revision,
+			});
+		}
+
 		for (const [field, fieldChange] of change) {
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts:647]: I would welcome a comment explaining what the point of this secon pass is.

File: packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts [hunk 6]
```diff
@@ -454,127 +461,296 @@ export class ModularChangeFamily
...
+		return rebasedChange;
 	}
 
 	public rebaseAnchors(anchors: AnchorSet, over: ModularChangeset): void {
 		anchors.applyDelta(this.intoDelta(over));
 	}
 
-	public intoDelta(change: ModularChangeset): Delta.Root {
-		return this.intoDeltaImpl(change.changes);
+	public intoDelta(change: ModularChangeset, repairStore?: ReadonlyRepairDataStore): Delta.Root {
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts:752]: Some leftover from merging Jenn's changes I'm guessing
```suggestion
	public intoDelta(change: ModularChangeset): Delta.Root {
```

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts [hunk 7]
```diff
@@ -145,6 +145,7 @@ function rebaseMarkList<TNodeChange>(
 				factory.push(cloneMark(currMark));
 			}
 		} else if (currMark === undefined) {
+			// TODO: Do we need to handle rebasing over baseMark's changes in this case?
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:148]: That does seem to be what the new contract requires.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts [hunk 8]
```diff
@@ -410,6 +411,11 @@ function rebaseMark<TNodeChange>(
 	moveEffects: MoveEffectTable<TNodeChange>,
 ): CellSpanningMark<TNodeChange> {
 	if (isSkipMark(baseMark) || isSkipLikeReattach(baseMark) || isSkipLikeDetach(baseMark)) {
+		// TODO: Rebase currMark's changes over baseMark's changes.
+		if (isObjMark(currMark) && currMark.type === "Modify") {
+			const childChange = rebaseChild(currMark.changes, undefined);
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:416]: Something like that, right?
```suggestion
			const childChange = rebaseChild(currMark.changes, isObjMark(baseMark) && baseMark.type !== "ReturnTo" ? baseMark.changes : undefined);
```

Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:416]: Although seeing how many similar cases need to be handled, it's probably best to make a function that takes any curr and base marks and calls rebaseChild with the appropriate child changes.

Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:416]: Agreed

File: packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts (hunk 1 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts:446]: Sounds good. A comment would be nice. :)

**Review by yann-achard-MS — APPROVED** [review]

File: packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts [hunk 9]
```diff
@@ -478,155 +490,334 @@ export class ModularChangeFamily
...
+		fieldsToAmend: Set<FieldChange>,
+		invalidatedEmptyFields: Set<FieldChange>,
+		genId: IdAllocator,
+		revisionMetadata: RevisionMetadataSource,
+		constraintState: ConstraintState,
+	) {
+		crossFieldTable.invalidatedFields = new Set();
+		for (const fieldToAmend of fieldsToAmend) {
+			const fieldContext = crossFieldTable.baseChangeToContext.get(fieldToAmend);
+			let newChange: FieldChange;
```
Comment (alex-pardes) [inline packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts:554]: Rename `newChange` to `rebasedChange`.

File: packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts [hunk 10]
```diff
@@ -478,155 +490,334 @@ export class ModularChangeFamily
...
+							newNodeChange === undefined && invalidatedEmptyFields.has(base),
+						revisionMetadata,
+						constraintState,
+					),
+				genId,
+				newCrossFieldManager(crossFieldTable),
+				revisionMetadata,
+			);
+
+			if (newNode !== undefined && field !== undefined) {
```
Comment (alex-pardes) [inline packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts:610]: This block could use a comment.

File: packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts [hunk 11]
```diff
@@ -478,155 +490,334 @@ export class ModularChangeFamily
...
+			change?.fieldChanges ?? new Map(),
+			baseMap,
+			genId,
+			crossFieldTable,
+			fieldFilter,
+			revisionMetadata,
+			constraintState,
+		);
+
+		const rebasedChange: NodeChangeset = {};
```
Comment (alex-pardes) [inline packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts:780]: We should return undefined here if `change === undefined && fieldChanges.size === 0`.

**Pull request merged**